lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -96,6 +97,24 @@ struct ConvertMemRefToCPURuntime
               }
             }
           }
+        } else if (isa<memref::AllocOp>(op)) {
```

Comment:
Please note that the `walk()` order may differ from the execution order. And what about the case

```
func AAA() {
     a=alloc()
     b=alloc()
     c=alloc()
     dealloc(b)
     dealloc(a)
     return c
}
```

The optimization still applies.

Reply:
Another topic is that, since we are going to handle general MLIR programs that may be hand-written by users, can your program handle this?

```
func AAA() {
     a=alloc()
     b=alloc()
     c=alloc()
     if(...) {
        dealloc(c)
     } else {
        dealloc(b)
     }
     dealloc(a)
}
```

It looks like stack-like, but it is not.

I would suggest a stricter checking:
1. first, check if an alloc is "leaked" from the scope (already done in your code in main branch)
2. for the non-leaked allocations only: check if the order of alloc/free is stack-like in the same **Region**, and ignore the leaked allocations. Usually, we can ensure the execution order in the same region.

Reply:
> Another topic is that, since we are going to handle general MLIR programs that may be hand-written by users, can your program handle this?
> 
> ```
> func AAA() {
>      a=alloc()
>      b=alloc()
>      c=alloc()
>      if(...) {
>         dealloc(c)
>      } else {
>         dealloc(b)
>      }
>      dealloc(a)
> }
> ```
Not sure if it's a valid case or not, as it will always result in memory leak?

Reply:
> > Another topic is that, since we are going to handle general MLIR programs that may be hand-written by users, can your program handle this?
> > ```
> > func AAA() {
> >      a=alloc()
> >      b=alloc()
> >      c=alloc()
> >      if(...) {
> >         dealloc(c)
> >      } else {
> >         dealloc(b)
> >      }
> >      dealloc(a)
> > }
> > ```
> 
> Not sure if it's a valid case or not, as it will always result in memory leak?

I think we are targeting a general MLIR compiler, so any of the user input should be handled. Users can always write code like this. At least it will not produce wrong result - if they can live with memory leak.